### ISSUE TITLE
sleigh: access to program counter

### ISSUE THREAD

**Issue by esaulenka** [NationalSecurityAgency/ghidra#1290]
I am writing sleigh module for Bosch/Infineon C166 controllers.
Now I faced with the following problem.

There is `EXTP` (extended page) instruction, that modifies addressing for next load/store instructions. `EXTP` has a parameter, that indicates, how much instructions are affected (any instructions are affected, not only dealing with memory ones).
Problem is complicated by the fact that C166 instructions have variable lenght, and `inst_next * const` not exactly matches end of EXTP action.

Are there any suggestions, how to solve this trouble?

**Comment by esaulenka** [comment]
I will continue this topic.

I found `delayslot` operator. 
With conjunction with context variables it should do what I need.

But, now I have another question: how to change context in constructor's body ?

```
# extended addressing fake register
define register offset=0x10 size=4	[ extAddressing ];
define context extAddressing
	extR = (31,31)
	extP = (30,30)
	extOffset = (0,23)
;
```

```
macro ext_p(page, irange) {
#	extP = 1; ### it not working !!
#	extOffset = page << 14;
	delayslot(irange);
#	extP = 0;
}
:extp #u1627, #irange is op0007=0xD7 & irange & op1415=1 ; u1627 & op2831=0 {
	ext_p(u1627, irange);
}
:extp r0811, #irange is op0007=0xDC & r0811 & irange & op1415=1 {
	ext_p(r0811, irange);
}

# carry extP flags in load/store instructions ...
```

Now I found one bad way, but it seems to be work:
`extp` set up couner (1, 2, 3, 4) in context variable with attribute `noflow` and any other instructions decrease it.

**Comment by caheckman** [comment]
Context variables can only be changed inside the "disassembly action" portion of a constructor (between the square braces '[' and ']'). They can be changed locally with a direct assignment or globally using the **globalset** directive.  See sections 8.2 "Local Context Change and 8.3 "Global Context Change" in the SLEIGH documentation.

**Comment by esaulenka** [comment]
Chris, thanks, I already found this info.
May be you will tell best way to model next behavior:
`extp` sets internal flag (I think, this flag fit to context register), execute N instructions (they may use this flag, but may be not) and then resets flag.

**Comment by rnd-ash** [comment]
@esaulenka I am trying to extend the C166 library for C167 processors, and are encountering the same issue here....The compiler Siemens uses for the C167 makes use of extr instructions continuously `extr #01`.

I was able to make extr instructions work via context changes to the extR fake register, but then that change is permanent.

Is there a way at all to make a context change again after an instruction is parsed?

**Comment by esaulenka** [comment]
@rnd-ash, I tried to use another approach [there](https://github.com/esaulenka/Ghidra_C166/commit/d494972c3bb0a9d91bd65dc486fcea63a4830bdc). Sorry, I forgot to push it after local tests.
In short, I added additional counter, that set by `extp #X` and decreases by all others instructions. If this counter is not zero, it changes the behavior of all memory access instructions.

But just few notes: first of all, this module is not complete, all memory accesses should be reworked somehow (now it produces weird code in most cases). And second one - I tested it on XC27xx MCU, which is successor of original C166/C167, so all C167 instructions should be already there.
And the last - I think will be better to discuss it on that particular project page instead of this common one.
